### ISSUE TITLE
`FFTMTF` normalization breaks autograd despite recent gradient-compatibility work

### ISSUE THREAD

**Issue by goldengrape** [optiland/optiland#571]
### Checklist

- [x ] I have searched the [existing issues](https://github.com/HarrisonKramer/optiland/issues) and [discussions](https://github.com/HarrisonKramer/optiland/discussions) for a similar question or feature request.
- [x ] I have read the [documentation](https://optiland.readthedocs.io/en/latest/) and tried to find an answer there.
- [x ] I am using the **latest version** of Optiland.
- [x ] I have included **all necessary context**.

*Thanks for taking the time to go through this — it really helps us help you!*

## Summary

Recent commits have made significant progress toward making the MTF module autograd-compatible, but one step in `FFTMTF._generate_mtf_data` silently breaks the computation graph: the normalization via `be.max()`. This makes `FFTMTF` values non-differentiable even when the PyTorch backend is active.

## Context: Recent Gradient-Compatibility Work

Two recent commits show that gradient support for the MTF module is an explicit goal:

- [**`2f19a60`**](https://github.com/optiland/optiland/commit/2f19a6024fafb7b27114479ae5c6dd4b4a98cb14) *(feat: update mtf module for torch backend compatibility)* — added `be.copy()` in `GeometricMTF._compute_field_data` with the comment `# copy required to maintain gradient`.
- [**`58bcfa4`**](https://github.com/optiland/optiland/commit/58bcfa4302438f1988b0cde71927017251ed90b0) *(refactor: replace shorthand addition with explicit addition for gradient tracking)* — replaced `FNO *= ...` with `FNO = FNO * ...` in `FFTMTF._get_fno` to preserve the computation graph.

These changes demonstrate clear intent to support autograd through the MTF pipeline. However, one step was missed.

## Root Cause

In `optiland/mtf.py`, `_generate_mtf_data` normalizes the MTF arrays using `be.max()`:

```python
# optiland/mtf.py, line 344
mtf.append([tangential / be.max(tangential), sagittal / be.max(sagittal)])
```

`be.max()` in the PyTorch backend is implemented as:

```python
# optiland/backend/torch_backend.py, lines 330–333
def max(x):
    if isinstance(x, torch.Tensor):
        return x.detach().cpu().max().item()
    return np.max(x)
```

The `.detach()` call explicitly removes the tensor from the computation graph, and `.item()` converts it to a Python scalar. Gradients cannot flow through this normalization step, so the entire `FFTMTF` output is non-differentiable — even though all upstream steps (FFT, complex multiplication, `fftshift`) are fully autograd-compatible in PyTorch.

## Expected Behavior

With the PyTorch backend and `grad_mode` enabled, `FFTMTF` values should carry gradients and be usable as loss functions in gradient-based optimization, consistent with `SpotDiagram.rms_spot_radius()` and `FFTPSF.strehl_ratio()`.

## Suggested Fix

Replace the `be.max()` call in the normalization step with a gradient-preserving operation. For example, introduce a `be.amax()` (or `be.differentiable_max()`) that dispatches to `torch.amax()` under PyTorch (which supports autograd) and `np.max()` under NumPy:

```python
# proposed change in _generate_mtf_data
mtf.append([tangential / be.amax(tangential), sagittal / be.amax(sagittal)])
```

This is a minimal, targeted fix consistent with the pattern already established in commits `2f19a60` and `58bcfa4`.

## Note on `GeometricMTF`

`GeometricMTF` has a separate, more fundamental limitation: it relies on `be.histogram()`, which is inherently non-differentiable (discrete binning). Fixing it would require a different approach (e.g., soft histogram or kernel density estimation) and is out of scope for this issue.

## References

- `optiland/mtf.py`, line 344 (`_generate_mtf_data`, normalization)
- `optiland/backend/torch_backend.py`, lines 330–333 (`be.max`)
- Commit [`2f19a60`](https://github.com/optiland/optiland/commit/2f19a6024fafb7b27114479ae5c6dd4b4a98cb14)
- Commit [`58bcfa4`](https://github.com/optiland/optiland/commit/58bcfa4302438f1988b0cde71927017251ed90b0)
- `docs/gallery/differentiable_ray_tracing/gradient_calculation.ipynb` (existing differentiable metrics example)

**Comment by RayPragma** [comment]
Hi — sharing what looks like good news for this issue: I believe this was already addressed (incidentally) by commit `db464edc` from PR #531 *"Refactor MTF Calculations for High-Accuracy Off-Axis Results"*.

That refactor replaced the `be.max()`-based normalization with **DC-bin normalization** (`data[center, center]`), which is physically more correct *and* autograd-safe, because indexing preserves the computation graph. The current code at [`optiland/mtf/fft.py:181-187`](https://github.com/optiland/optiland/blob/master/optiland/mtf/fft.py#L181-L187) reads:

```python
dc_value = data[center, center]
if dc_value == 0:
    norm_tangential = be.zeros_like(tangential)
    norm_sagittal = be.zeros_like(sagittal)
else:
    norm_tangential = tangential / dc_value
    norm_sagittal = sagittal / dc_value
```

No `be.max()` is involved in `FFTMTF._generate_mtf_data` any more.

### End-to-end verification

I ran a quick repro (single thickness leaf tensor → backward through `FFTMTF.mtf`):

```python
import torch
import optiland.backend as be
from optiland import optic
from optiland.mtf import FFTMTF

be.set_backend("torch"); be.grad_mode.enable()

lens = optic.Optic()
lens.surfaces.add(index=0, thickness=be.inf)
lens.surfaces.add(index=1, thickness=7, radius=1000, material="N-SF11", is_stop=True)
lens.surfaces.add(index=2, thickness=30, radius=-1000)
lens.surfaces.add(index=3)
lens.set_aperture(aperture_type="EPD", value=15)
lens.fields.set_type(field_type="angle"); lens.fields.add(y=0)
lens.wavelengths.add(value=0.55, is_primary=True)

t2 = torch.tensor(30.0, dtype=torch.float64, requires_grad=True)
lens.updater.set_thickness(t2, surface_number=2)

mtf = FFTMTF(lens, fields="all", num_rays=32)
mtf.mtf[0][0].sum().backward()
print(t2.grad.item())  # -0.00012207 (non-zero, finite, signed — autograd flows end-to-end)
```

So unless I'm missing a code path the original report had in mind, this issue can be closed.

> ### Tangentially related (probably a separate issue)
> 
> `be.max()` *still* breaks autograd in [`optiland/psf/mmdft.py:221`](https://github.com/optiland/optiland/blob/master/optiland/psf/mmdft.py#L221) (`FFTPSF.strehl_ratio`), which returns a Python `float` rather than a tensor. Same root cause (`be.max` in the torch backend calls `.detach().cpu().max().item()`), different metric. Happy to file that as its own issue if useful.

Hope this helps!
